New feature RobustWeightedEstimator : code + examples

TimotheeMathieu opened the pull request:
Algorithms to do robust classification and regression.

Example of use of this code gives the two attached plots (the code used to generate the plots is in the example folder). The principle is to have algorithms that are robust to outliers both in the feature X and in the labels Y. Most sklearn algorithm allow robust loss functions that exhibit only the robustness in Y (except for Ransac and Theilsen estimators).

![sklearn-extra_robust_classif](https://user-images.githubusercontent.com/30346931/67620091-5b306780-f803-11e9-8df6-089bf07ad998.png)
![sklearn-extra_robust_regression](https://user-images.githubusercontent.com/30346931/67620092-5b306780-f803-11e9-9ffb-6a94b78e46ae.png)

jnothman commented on sklearn_extra/robust/RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,362 @@
+    This model use iterative reweighting of samples to make a regression or
+    classification estimator robust.
+
+    This algorithm is stille very new and still in development.
```
This does not meet our criteria for inclusion. We cannot afford to include and maintain new algorithms, and recommend you publish your implementation elsewhere.

jnothman replied:
I take it back. I didn't see this pull request was in contrib. All good!

rth commented:
The codecov CI failure can be fixed by adding tests to increase the code coverage of added code..

TimotheeMathieu commented on sklearn_extra/robust/RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,362 @@
+    ----------
+
+    [1] Guillaume Lecué, Matthieu Lerasle and Timothée Mathieu.
+        "Robust classification via MOM minimization", arXiv preprint (2019).
```
Yes, but in 2019 the article have been accepted for publication in Springer Machine Learning journal... when the publication is finalized I will change the citations.

TimotheeMathieu commented on sklearn_extra/robust/RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,362 @@
+    This algorithm enjoy a non-zero breakdown-point (it can handle arbitrarily
+    bad outliers). When the "mom" weighting scheme is used, K/2 outliers can be
+    tolerated. When the "Huber" weighting scheme is used, asymptotically the
+    number of outliers has to be less than half the sample size.
```
There is no definite definition of what is an outlier. Here I just say that we follow the majority, if a minority of the points are very different from the general behavior, then we bound their influence so that they don't have too much of an effect on the result of the prediction.
We know from theory that we can't handle more than half the data in "outliers", it is rather intuitive, but more than that this limit can be attained (think about the median) and I only say that asymptotically the Huber weighting scheme give me an estimator that is close to be as robust as a median estimator, but the point is that it is more efficient than the median. 
I will clarify.

TimotheeMathieu commented on sklearn_extra/robust/RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,362 @@
+        If None, c is estimated at each step using half the Inter-quartile
+        range.
+
+    K : int, optional (default=3)
```
I don't understand, what do you propose then ? The number of outliers is unknown in general and most of the time we want K to be as small as possible while staying above the number of outliers, it is why I put the 3 (the minimal value) here, we may want to estimate the number of outliers to estimate K but in my experience cross-validation is the best way to choose K.

TimotheeMathieu replied:
On the other hand, you are right that the sample_size should be mentionned and should also intervene in the validation. Fixed and I also changed the definition of k a bit so that we want k to be about the number of outliers, it is simpler.

TimotheeMathieu commented on sklearn_extra/robust/RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,362 @@
+
+        if self.loss in ["log", "hinge", "squared_hinge"]:
+            # If in a classification task, precise the classes.
+            base_estimator.partial_fit(X, y, classes=list(set(y)))
```
because partial_fit ask for it, you can't do partial fit if you don't pass the classes explicitely (an error is raised). Maybe I should use fit with max_iter=1 ?

rth replied:
> because partial_fit ask for it, you can't do partial fit if you don't pass the classes explicitely (an error is raised). Maybe I should use fit with max_iter=1 ?

Makes sense. Let's pre-compute classes as `np.unique(y)` once outside the loop and re-use it here then.

chkoar commented:
I would use snake_case in filenames.

chkoar commented on doc/modules/robust.rst:
```diff
@@ -0,0 +1,246 @@
+influence that any minority of the dataset can have on the prediction.
+
+.. |robust_regression| image:: ../robust_plot_regression.png
+       :target: ../examples/plot_RobustRegression.py
```
You could rename the original file like `plot_robust_regression.py`

chkoar commented on examples/plot_RobustRegression_california_houses.py:
```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
+"""
+================================================================
+A demo of Robust Regression on Real dataset "california housing"
```
The `real` should be in lower case or am I missing something?

TimotheeMathieu commented on sklearn_extra/robust/RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,362 @@
+    tolerated. When the "Huber" weighting scheme is used, asymptotically the
+    number of outliers has to be less than half the sample size.
+
+    Read more in the :ref:`User Guide <RobustWeightedEstimator>`.
```
I made a user guide file but I am not sure how to make a reference to it in the code, and I am also not sure how to compile the doc in local, can it be done ? For now, I did the rst doc file without checking the details because I could not compile it locally except with rst2html.

chkoar commented on examples/plot_robust_regression_toy.py:
```diff
@@ -0,0 +1,95 @@
+    RANSACRegressor,
+    HuberRegressor,
+)
+
```
```python 
from sklearn.utils import shuffle
```

chkoar commented on examples/plot_robust_regression_toy.py:
```diff
@@ -0,0 +1,95 @@
+# Shuffle the data so that we don't know where the outlier is.
+perm = rng.permutation(len(X))
+X = X[perm]
+y = y[perm]
```
```python
X, y = shuffle(X, y, ranndom_state=rng)
```

chkoar commented on doc/modules/robust.rst:
```diff
@@ -0,0 +1,246 @@
+We use the term "outlier" to mean a discordant minority of the dataset and it
+can comes in a lot of different forms. Most usual are outliers that are
+situated outside the bulk of the data. They can can be outliers in the features
+X or in the labels Y (see the plots below).
```
I am not a native speaker but it needs some polishing in the expressions. Probably and the rest of the doc.

TimotheeMathieu replied:
I am not a native speaker either so I can't promise anything, but you are right there were a lot of typos and I corrected some.

TimotheeMathieu replied:
I changed some things. 
I am open to further suggestions but for now I don't see any big problems in the docs.

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+
+    Attributes
+    ----------
+    estimator : object,
```
estimator_ : object
        The fitted base estimator.

chkoar replied:
The method take as parameter the `base_estimator` but it stores the estimator in the `estimator_`. You should use the same name for consistency.

TimotheeMathieu replied:
I used the same convention as RansacRegressor. Should I use base_estimator_ ?

chkoar replied:
IMHO we should use what is the most common convention that is used in the `scikit-learn` repo. So, the parameter and the attribute will have the same name. Personally I do not have a strong opinion but the most common is the `base_estimator`, as I can see.

chkoar commented on sklearn_extra/robust/tests/test_RobustWeightedEstimator.py:
```diff
@@ -0,0 +1,129 @@
+import numpy as np
```
Probably `test_robust_weighted_estimator.py`

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+        if self.loss_ is None:
+            self.loss_ = "squared_loss"
+
+        base_estimator.set_params(
```
You make the assumption that the user will pass a `SGDRegressor`  or `SGDClassifier` instance but you should not. IMHO I think that you need to validate in fit (as usually in `scikti-learn`) and raise a helpful message.

TimotheeMathieu replied:
I resolved this. 
But on the other hand, there is only SGDRegressor and SGDClassifier that have what I want already coded in scikit-learn. Other algorithms could be used but they have not sample_weight or partial_fit implemented. The idea would be for the user to code the algorithm himself but for that he would have to understand all the things his algorithm need to verify. I don't know how I could make it easier for the user to use anything other than SGDRegressor and SGDClassifier.

chkoar replied:
Sorry, I didn't get into the details of your method.

> Other algorithms could be used but they have not sample_weight or partial_fit implemented.

I think that most of the predictors support `sample_weight`.  Apart from that there are [several](https://scikit-learn.org/0.15/modules/scaling_strategies.html#incremental-learning) methods that support `partial_fit`. So, if your intention is to work with estimators that support `partial_fit` and/or `sample_weight` you could decouple the specific SGD parameters and let the user pass his/her estimator. Then, in your `fit` method you could validate if the passing estimator has the required parameters (like `partial_fit` and `sample_weight` in fit signature)

TimotheeMathieu replied:
In fact, most estimators don't support sample_weight. For example `MiniBatchKMeans` would work but it does not support sample_weight. Similarly `MLPClassifier/Regressor` would work but also don't support sample_weight. And although `BernoulliNB` supports partial_fit but it does not have a warm_start parameter...
In there, I say that I only support `SGDRegressor/Classifier` because it made things easier for me, I could refer to specific names of parameters from these two algorithms like `eta0` and `learning_rate`. This type of parameter that control the optimization process is common to a lot of algorithm and I don't need them per-se it only makes the convergence better.
I fear that if I work to allow more general estimators, I may need to loose in efficiency in the process.

TimotheeMathieu replied:
I was wrong, in fact MiniBatchKmeans support sample_weight it is just not in the documentation, but if one look at the code it is implemented. 
I will make an example with Kmeans to illustrate how one can use any algorithm that support sample_weight and partial_fit only.

TimotheeMathieu replied:
The example is rather pretty and easy to code so it is a good idea. Thank you for suggesting MiniBatchKMeans I had only looked at KMeans.

![robust_kmeans](https://user-images.githubusercontent.com/30346931/68400355-311e6400-0178-11ea-8ceb-c602454a4581.png)

I think I will include this. It only change the code a little to be able to handle unsupervised learning as well as supervised learning.

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+
+        self.c_ = self.c
+        self.k_ = self.k
+        self.loss_ = self.loss
```
Do we need to store these attributes?

TimotheeMathieu replied:
I need to be able to change the parameter, hence the underscore. I don't really know what is the convention in the case of a parameter that later changes.

chkoar replied:
Read [hear](https://scikit-learn.org/stable/glossary.html#term-attributes) to understand what is the purpose of the learn (and not) attributes during fit.

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+        estimator trained using the RobustWeightedEstimator scheme, can be used
+        as any sklearn estimator.
+
+    weights : array like, length = n_sample.
```
`weights_`

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+            perm = random_state.permutation(len(X))
+            X = X[perm]
+            y = y[perm]
+            weights = weights[perm]
```
Couldn't we replace these 4 lines with the following?
```python
X, y, weights = shuffle(X, y, weights)

```

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+
+        if self.max_iter <= 0:
+            raise ValueError(
+                "RobustWeightedEstimator: "
```
IMHO `RobustWeightedEstimator` is unnecessary. Otherwise you can user `self.__class__.__name__`.

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+
+        if self.burn_in < 0:
+            raise ValueError(
+                "RobustWeightedEstimator: "
```
Same as above

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+
+        if (self.burn_in > 0) and (self.eta0 <= 0):
+            raise ValueError(
+                "RobustWeightedEstimator: "
```
Same as above

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+            or self.k > np.floor(n / 2)
+        ):
+            raise ValueError(
+                "RobustWeightedEstimator: "
```
Same as above

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+                self.c_ = iqr(np.abs(loss_values - np.median(loss_values))) / 2
+                if self.c_ == 0:
+                    warnings.warn(
+                        "RobustWeightedEstimator: "
```
Same as above

chkoar commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,442 @@
+        self.loss = loss
+        self.max_iter = max_iter
+        self.random_state = random_state
+
```
Since you use the same class for classification and regression problems you need some where to set the estimator type

```python
    @property
    def _estimator_type(self):
        return self.estimator._estimator_type
```

chkoar replied:
The same goes for the `score` method of the estimator. Take a look [here](https://github.com/scikit-learn/scikit-learn/blob/452d919e42b16771701a8043f9807597df279041/sklearn/model_selection/_search.py#L417).

TimotheeMathieu commented:
Thank you both for the reviews @rth and @chkoar 

- I added a manual section I hope it makes things easier to understand. I was not able to compile the doc locally so I am not sure if the references to figures works.
- I did not do the Benchmark, @rth  I am not sure what you mean by "Benchmark" as the computation time would be different on different computer and as this algorithm is not particularly fast I can't really compare the computation time with other algorithms except to say that it is slower.
- I did 2 examples on real datasets, but I was not sure how to do it because in order to assert the efficiency of the method, I had to do some validation steps similar to cross-validation and this is computationally costly and I didn't want to have examples that take a lot of time to run. I did a compromise by doing validation but with less steps that what I usually use the evalutation is less precise but at least it should avoid overfit.
- I also added an example with KMeans, as this illustrate the use of robust unsupervised learning but also of how to adapt an estimator so that it runs with this meta-algorithm.

TimotheeMathieu commented:
The CI is failed because of _fastfood.py that does not respect black format. 
Should I open an issue for that or can I just reformat in my code to have all the CI's successful ?

chkoar commented:
IMHO that need a separate PR

rth commented on examples/plot_robust_regression_california_houses.py:
```diff
@@ -0,0 +1,110 @@
+from sklearn.preprocessing import RobustScaler
+
+
+def eval(est, X, y, X_test, y_test):
```
`eval` is a python keyword, better to name it something else..

TimotheeMathieu commented on doc/modules/robust.rst:
```diff
@@ -0,0 +1,251 @@
+
+Outliers can be caused by a lot of things, among them are human errors, captor errors or inherent causes.
+For example one can think in an economical setting to income which is highly
+corrupted by outliers (there are people with very high income).
```
Well I know that this is typically the type of data where people use trimmed mean or median or things like that, robust methods. In my opinion, this is typically the type of example where one might want to use robust methods instead of outlier detection methods and then delete the outliers. The principle is really that these are outliers but we want them to be of influence nonetheless, we just want the influence to be bounded. In econometrics, we can use for example quantile regression and this is a type of robust method.
The duality outlier/heavy tail is difficult to understand I think, but they both stem from the same source. Robust methods are, depending on the person you ask, methods that can deal with outliers or methods that deals with heavy tail data (typically we want to be as efficient on Gaussian data as on data with only a finite variance but no third moment).

You are right in the sense that maybe I should not call this outliers but heavy tailed data but the robust methods are what need to be used in that case in my opinion.

TimotheeMathieu commented on doc/modules/robust.rst:
```diff
@@ -0,0 +1,251 @@
+
+.. centered:: |outlier|
+
+Outliers can be caused by a lot of things, among them are human errors, captor errors or inherent causes.
```
I wanted to refer to something like the income for example (see the other comment).

TimotheeMathieu commented:
I updated the doc, I had someone review it a little. If you feel that some parts are not needed (is it interesting to put a little theory ?) do not hesitate to tell me.

> I meant having a rough estimate of the time/memory complexity. Say of a dataset of shape `(n_samples, n_features)` with the SGDClassifier, is it going to a few times slower independently of the dataset size, or is it going to be slower by some `O(n_samples**p * n_features**m)` in which case for high exponents it would be difficult to use this approach on very large datasets; i.e is it suitable for large datasets or are there known limitations.

We have a theoretical expression of the time added, in theory it is in O(n_sample*log(n_sample)+n_sample*L(n_feature)) where L(n_feature) is the complexity of evaluating the loss function, in most cases this is linear. From this, this algorithm is suitable for high dimension application. I tried the algorithm on openml "arcene" dataset (classification with n_sample=200, n_features=10 000). The result was that SGDClassifier runs a 10 fold cross-validation in 1s and RobustWeightedEstimator(SGDClassifier()) runs the same 10 folds cross-validations in 13.2s for approximately the same efficiency (I didn't tune anything in both algo). 
What I don't really see is how to make a script that use this.

TimotheeMathieu commented:
I added what could be a benchmark plot, it plots the time for classification on a dataset with increasing  number of features and increasing number of samples. The plot seems linear, it fits the theory.

chkoar commented:
@TimotheeMathieu can you fix the failing tests?

TimotheeMathieu commented:
The problem comes from EigenPro and as it is not my code I could not find quickly where is the bug.

chkoar commented:
I saw [this](https://travis-ci.org/scikit-learn-contrib/scikit-learn-extra/jobs/627615800#L853) that's why I asked.

TimotheeMathieu commented:
The link you gave is of the test 136.1, the current Travis test is [136.4](https://travis-ci.org/scikit-learn-contrib/scikit-learn-extra/jobs/627615803?utm_medium=notification&utm_source=github_status) from what I gathered (I am new to CI tests).

chkoar commented:
Basically is one [build](https://travis-ci.org/scikit-learn-contrib/scikit-learn-extra/builds/627615799?utm_source=github_status&utm_medium=notification) with different configurations. The one you are referring to is against the nightly version of the `scikit-learn`. For instance the two first builds are against `0.21.2` version of `scikit-learn` where the robust estimator doc test fails due to changes to `check_is_fitted`. 

@rth whats your opinion here? In the README we say that we support scikit-learn (>=0.21)

TimotheeMathieu commented:
Ok it is fixed, the problem of compatibility was with sklearn0.21, in my version (sklearn 0.22) there was no problem so I didn't see it. Thanks for explaining how CI work.

rth commented:
@ngoix What's your opinion about this general approach outlined in the user guide? also cc @agramfort if you have any opinions.

GaelVaroquaux commented:
> Sorry this is taking a long time to merge. I was hoping that scikit-learn-extra would be faster there than scikit-learn, otherwise it defeats the purpose of this repo somewhat.

The bottleneck is the same: the reviewer time.

TimotheeMathieu commented:
@rth  No worries, I am already thankful that you all helped me to make it better and that you accepted to teach a rookie how to do a PR and we can't expect you to be a lot faster than scikit-learn when you have in fact less manpower.

rth commented:
@TimotheeMathieu Could you please merge upstream/master into this branch, that should fix the failing CI.

TimotheeMathieu commented:
I merged but CI still fail, it is only for python3.8 and seems to be because of sklearn-extra/utils/_cyfht.pyx I don't understand the import of pyx files so I can't help. I have python38 and the test also fail for master.

rth commented on sklearn_extra/robust/robust_weighted_estimator.py:
```diff
@@ -0,0 +1,486 @@
+                raise ValueError("The loss %s is not supported. " % self.loss)
+
+            loss_class, args = eff_loss[0], eff_loss[1:]
+            return np.vectorize(loss_class(*args).dloss)
```
The attribute name here is different depending on the scikit-learn version  due to https://github.com/scikit-learn/scikit-learn/pull/17191,
```suggestion
			  # at the top of the file
			  from distutils.version import LooseVersion
			
			  if LooseVersion(sklearn.__version__) >= LooseVersion("0.24.0"):
			      dloss_attr = 'py_dloss'
			  else:
			      dloss_attr = 'dloss'
			
            return np.vectorize(geattr(loss_class(*args), dloss_attr))
```

that should fix some error in the Python 3.8 CI.

TimotheeMathieu commented:
LooseVersion(sklearn.__version__) does not work in python3.8 with scikit-learn dev version, it does not parse current master of sklearn which is '0.24.dev0' (this contain a string) and this implies that your fix does not work, I don't know how to cleanly fix this but we could do a custom parser like 

`if sklearn.__version__.split('.')[1]>=24:`

would this be acceptable ? It is a little hackish.

rth commented:
Yes, looks like `LooseVersion` is actually not reliable. I pushed a commit to switch to `pkg_resources.parse_version` , where `pkg_ressources` is installed as part of `setuptools`. It would mean making setuptools a run time dependency in addition to being a build dependency. There doesn't seem to be much better solution for this https://github.com/scikit-learn/scikit-learn/issues/7980#issuecomment-645357824

rth commented:
Merged. Thank you @TimotheeMathieu !
